## Request example

[(ref)](https://realpython.com/api-integration-in-python/)
To test out GET and the other methods in this section, we use a service called JSONPlaceholder. This free service provides fake API endpoints that send back responses that requests can process.

In [4]:
import requests as req

### GET
Retrieving resources from a given API.

In [4]:
api_url = "https://jsonplaceholder.typicode.com/todos/1"
response = req.get(api_url)
response.json()

{'userId': 1, 'id': 1, 'title': 'delectus aut autem', 'completed': False}

This code calls requests.get() to send a GET request to /todos/1, which responds with the todo item with the ID 1. Then you can call .json() on the response object to view the data that came back from the API.

We can check HTTP status code which provides the information about the request. For example,
**200** means "The requested action was successful".


In [7]:
response.status_code 

200

### POST
Posting data to a REST API to create a new resource. 

In [9]:
api_url = "https://jsonplaceholder.typicode.com/todos"
todo = {"userId": 87, "title": "ぺこみこ最強", "completed": False}
response = req.post(api_url, json=todo)
response.json()

{'userId': 87, 'title': 'ぺこみこ最強', 'completed': False, 'id': 201}

Here, we call requests.post() to create a new todo in the system.
First, you create a dictionary containing the data for your todo. Then you pass this dictionary to the json keyword argument of requests.post(). When you do this, requests.post() automatically sets the request’s HTTP header Content-Type to application/json. It also serializes todo into a JSON string, which it appends to the body of the request. Since it is only virtual endpoint for simulation, no data will really be created.

In [11]:
response.status_code 

201

### PUT
Updating an existing todo with new data.

In [19]:
api_url = "https://jsonplaceholder.typicode.com/todos/10"
response = req.get(api_url)
response.json() #Our target to be updated

{'userId': 1,
 'id': 10,
 'title': 'illo est ratione doloremque quia maiores aut',
 'completed': True}

In [23]:
todo = {"userId": 1, "title": "幹你媽去死", "completed": True} #update content
response = req.put(api_url, json=todo)
response.json()

{'title': '幹你媽去死', 'id': 10}

Here, you first call requests.get() to view the contents of the existing todo. Next, you call requests.put() with new JSON data to replace the existing to-do’s values. You can see the new values when you call response.json(). Successful PUT requests will always return 200 instead of 201 because you aren’t creating a new resource but just updating an existing one.

In [22]:
response.status_code

200

### PATCH
Modifying(partially updating) an existing todo with new data (it doesn’t completely replace the existing resource).

In [25]:
api_url = "https://jsonplaceholder.typicode.com/todos/10" 
todo = {"title": "幹你媽去死"} #only change title
response = req.patch(api_url, json=todo)
response.json()

{'userId': 1, 'id': 10, 'title': '幹你媽去死', 'completed': True}

In [26]:
response.status_code

200